### PR TITLE
shell: live output

### PR DESCRIPTION
**Pull request by chu11**
builds on top of my misc cleanups PR in #2360.

- have shell send output to KVS live as it happens
- have shell emit a "output-ready" event to `exec.eventlog`
  - good name?  I considered "output {"ready":"true"}" as well.
- job attach checks for "output-ready" from `exec.eventlog`, and then reads / outputs from `guest.output` as it occurs.
- Fix racy completion in shell as discussued in #2363

### CODE REVIEW

File: src/shell/output.c [hunk 1]
```diff
@@ -94,9 +95,87 @@ static void shell_output_control (struct shell_output *out, bool stop)
...
+{
+    struct shell_output *out = arg;
+
+    if (flux_future_get (f, NULL) < 0)
+        log_err ("shell_output_commit");
+    zlist_remove (out->pending_commits, f);
+    flux_future_destroy (f);
+}
+
+static int shell_output_commit (struct shell_output *out, json_t *entry)
```
Comment (garlick) [inline src/shell/output.c]: This code should probably not be limited to one eventlog entry per KVS commit, since that's just one extreme of what we will likely need to support.  I would suggest continuing to keep a list of entries in `struct shell_output`, and write this function so it aggregates all of them in one KVS transaction and empties the list.  For the initial implementation, we'll just call this function after every append, but later we can call it based on elapsed time and/or number of entries in the list.

File: src/shell/output.c [hunk 2]
```diff
@@ -292,15 +307,21 @@ static int shell_output_header_append (flux_shell_t *shell, json_t *output)
...
                              "options");
-    if (!o || json_array_append_new (output, o) < 0) {
-        json_decref (o);
+    if (!o) {
         errno = ENOMEM;
-        return -1;
+        goto error;
     }
-    return 0;
+    if (shell_output_commit (out, o) < 0)
```
Comment (garlick) [inline src/shell/output.c]: Now that we have the guest eventlog event, header can go out in the same commit as the first data entry.

File: src/shell/output.c [hunk 3]
```diff
@@ -78,6 +81,65 @@ static void shell_output_control_task (struct shell_task *task,
...
+    }
+    if (!(txn = flux_kvs_txn_create ())) {
+        log_err ("flux_kvs_txn_create");
+        goto error;
+    }
+    if (flux_kvs_txn_put (txn, FLUX_KVS_APPEND, key, entrystr) < 0) {
+        log_err ("flux_kvs_txn_put");
+        goto error;
+    }
+    if (!(out->output_ready_f = flux_kvs_commit (out->shell->h,
```
Comment (garlick) [inline src/shell/output.c]: This should go in the same commit as the output header (and probably the first data entry).

**Comment by grondo** [comment]
I think we can accept the bug for now and I'll be sure it gets fixed in my PR

**Comment by chu11** [comment]
rebased

File: src/shell/output.c [hunk 4]
```diff
@@ -13,14 +13,18 @@
...
  * selected I/O mode.
  *
- * The leader shell implements an "shell-<id>.output" service that
- * all ranks send task output to.  Output objects accumulate in a json
- * array on the leader.  Upon task exit, the array is written to the
- * "output" key in the job's guest KVS namespace.
+ * The leader shell implements an "shell-<id>.output" service that all
+ * ranks send task output to.  Output objects accumulate in a json
+ * array on the leader.  Output objects are written to the "output"
+ * key in the job's guest KVS namespace, or directly to stdout/stderr
```
Comment (garlick) [inline src/shell/output.c]: Cite RFC 24?

File: src/shell/output.c [hunk 5]
```diff
@@ -325,7 +379,9 @@ struct shell_output *shell_output_create (flux_shell_t *shell)
             errno = ENOMEM;
             goto error;
         }
-        if (shell_output_header_append (shell, out->output) < 0)
+        if (!(out->pending_commits = zlist_new ()))
+            goto error;
+        if (shell_output_header (out) < 0)
```
Comment (garlick) [inline src/shell/output.c]: You mentioned you wanted failure to commit the header to be fatal, but it looks like if the commit RPC returns an error, it is non-fatal in `shell_output_commit_completion()`?

File: src/shell/output.c [hunk 6]
```diff
@@ -204,6 +207,10 @@ static int shell_output_commit (struct shell_output *out)
         out->output_ready_sent = true;
     if (flux_future_then (f, -1, shell_output_commit_completion, out) < 0)
         goto error;
+    if (flux_shell_add_completion_ref (out->shell, "output.commit") < 0) {
```
Comment (garlick) [inline src/shell/output.c]: With the completion references taken for outstanding futures, will `shell_output_destroy()` ever be called with a non-empty `out->pending_writes`? (Maybe tracking futures in the list is not necessary anymore?)

File: src/shell/output.c (hunk 5 above)
Comment (chu11) [reply src/shell/output.c]: ahh good catch.  I was primarily concerned with fatal error on the request side, had forgotten the response side.

Comment (chu11) [reply src/shell/output.c]: hmmm, on such an error would it be better to run `flux_shell_killall()` rather than just exit?  B/c all tasks have started.

File: src/shell/output.c (hunk 6 above)
Comment (chu11) [reply src/shell/output.c]: I actually had pondered that, and thought there were still circumstances it was needed, such as if the shell was signaled.  But now that I look at it again, perhaps its not needed afterall.  I'll remove.

File: src/shell/output.c (hunk 5 above)
Comment (grondo) [reply src/shell/output.c]: I think we need a `shell_fatal_error ()` or similar call that attempts to clean up before a fatal error when tasks have started, instead of issuing the killall at the site of the error. Or perhaps with some thinking we could come up with a cleaner way to handle fatal errors with proper cleanup -- haven't thought about it too much at this point.

Comment (chu11) [reply src/shell/output.c]: yeah, I was just starting to think that.  Perhaps for now, `log_err_exit()` is fine and we can figure out the proper / best / cleanest way to deal with this in another issue.

**Comment by chu11** [comment]
rebased & re-pushed, addressing @garlick's comments above.  Per some discussion above since there isn't a very clean way to "fatal error" after tasks have started, I lazily call `log_err_exit()` on the fatal error of the output header / output-ready not succeeding.  Created a new issue #2378 for this particular case.

Removed the "pending_commits" list, as it was no longer necessary./

Found a corner case along the way, we shouldn't call `shell_output_commit()` for the output header in the standalone case.  We should call `shell_output_flush()` instead.  Technically, `shell_output_flush()` doesn't use the header at the moment, but that will change in the future.

**Comment by garlick** [comment]
Having another look at this.  It looks pretty close IMHO.

One thing I'm not too sure about is the decision to make a failed KVS commit after the header non-fatal.  Losing output could be a serious thing, depending on how the job is using its I/O.  My vote would be to make all related (unlikely) failures fatal to the shell to be sure the job fails.   When we get a better interface that cleans up any tasks, we can switch to that, but I worry a bit about allowing a job to report success it's in fact misplaced the "42" that was the result of your long simulation (OK probably an extreme hypothetical!)

**Comment by chu11** [comment]
Re-based & re-pushed with suggested fix of making errors on commit (and I also did flushes to terminal) fatal and calling `log_err_exit()`.  Very tiny fixes compared to what you just reviewed.

```
diff --git a/src/shell/output.c b/src/shell/output.c
index 7d82adc..a2229e5 100644
--- a/src/shell/output.c
+++ b/src/shell/output.c
@@ -141,13 +141,10 @@ static void shell_output_commit_completion (flux_future_t *f, void *arg)
 {
     struct shell_output *out = arg;
 
-    if (flux_future_get (f, NULL) < 0) {
-        if (flux_future_aux_get (f, "output::ready"))
-            /* failng to commit output-ready or header is a fatal
-             * error.  Should be cleaner in future. Issue #2378 */
-            log_err_exit ("shell_output_commit ready");
-        log_err ("shell_output_commit");
-    }
+    /* Error failing to commit is a fatal error.  Should be cleaner in
+     * future. Issue #2378 */
+    if (flux_future_get (f, NULL) < 0)
+        log_err_exit ("shell_output_commit");
     flux_future_destroy (f);
 
     if (flux_shell_remove_completion_ref (out->shell, "output.commit") < 0)
@@ -206,12 +203,8 @@ static int shell_output_commit (struct shell_output *out)
     }
     if (!(f = flux_kvs_commit (out->shell->h, NULL, 0, txn)))
         goto error;
-    if (!out->output_ready_sent) {
-        /* used as flag, content of set irrelevant */
-        if (flux_future_aux_set (f, "output::ready", f, NULL) < 0)
-            goto error;
+    if (!out->output_ready_sent)
         out->output_ready_sent = true;
-    }
     if (flux_future_then (f, -1, shell_output_commit_completion, out) < 0)
         goto error;
     if (flux_shell_add_completion_ref (out->shell, "output.commit") < 0) {
@@ -261,13 +254,15 @@ static void shell_output_write_cb (flux_t *h,
         errno = ENOMEM;
         goto error;
     }
+    /* Error failing to commit is a fatal error.  Should be cleaner in
+     * future. Issue #2378 */
     if (out->shell->standalone) {
         if (shell_output_flush (out) < 0)
-            log_err ("shell_output_flush");
+            log_err_exit ("shell_output_flush");
     }
     else {
         if (shell_output_commit (out) < 0)
-            log_err ("shell_output_commit");
+            log_err_exit ("shell_output_commit");
     }
     if (eof) {
         if (--out->eof_pending == 0) {

```

**Comment by grondo** [comment]
I accidentally ran an executable that didn't exist in this branch and it seems to cause a hang from which you can't escape:
```console
$ flux srun nofile
flux-job: task(s) exited with exit code 1
^Cflux-job: one more ctrl-C within 2s to cancel or ctrl-Z to detach
^Cflux-job: cancel: unknown job id
^Cflux-job: cancel: unknown job id
^Cflux-job: cancel: unknown job id
^Cflux-job: cancel: unknown job id
```

This doesn't occur on master:
```console
$ flux srun nofile
flux-job: task(s) exited with exit code 1
$
```

**Comment by chu11** [comment]
@grondo aw crud.  I'm thinking there's a corner case there where the guest.exec.eventlog will never be created, thus a hang forever.  The hang doesn't exist in my branch with this and the fix for #2338 (PR #2374).

Minimally a test should be added b/c we don't have a test for this.  Think we should combine PR #2374 with this issue and make a test?  Or try to fix in this PR?  Or just deal with it in that PR?  I'll see how easy/hard it might be to fix.

**Comment by garlick** [comment]
Is it possible to reverse the order of this and PR #2374?

**Comment by chu11** [comment]
Basically in this case, the job has "started", but has an obvious error which leads to no output.  `flux-job` sees that job has "started" and believes there is output, so begins watching `guest.output`.  No EOFs ever come, thus the watch on `guest.output` hangs.

Potential solutions:
- #2374 - the `guest.output` will return ENODATA at the of `guest.output`
- output the `output-ready` event later, after the first "real" stdout/stderr occurs, or perhaps when the state in `libsubprocess` turns to RUNNING, so we know the subprocess actually ran for real.

Re-ordering #2374 first is totally doable/fine and would solve this, but I think the latter solution is more correct.  I was previously outputting "output-ready" when we output the header (under the assumption that meant the job was running).  But that no longer seems correct.

**Comment by grondo** [comment]
> output the output-ready event later, after the first "real" stdout/stderr occurs

For a job with little or no output, would this cause `output-ready` event to be delayed until the first output stream is closed? If the `output-ready` event is delayed until the first other stream event, then is there truly a need for the `output-ready` event?

BTW, I was actually surprised there was no output at all in this case. Even for an exec failure, shouldn't the child processes still generate some stderr of the form `exec: No such file or directory` or similar?

Plugins that run in the pre-exec hook may also generate errors to stderr, which we have to ensure the shell captures and sends to the output eventlog, even if the error is fatal.

**Comment by chu11** [comment]
> For a job with little or no output, would this cause output-ready event to be delayed until the first output stream is closed? If the output-ready event is delayed until the first other stream event, then is there truly a need for the output-ready event?

Yeah.  Outputting `output-ready` when the job is in the RUNNING state is better/correct.

> BTW, I was actually surprised there was no output at all in this case. Even for an exec failure, 
> shouldn't the child processes still generate some stderr of the form exec: No such file or directory > or similar?
>
> Plugins that run in the pre-exec hook may also generate errors to stderr, which we have to ensure > the shell captures and sends to the output eventlog, even if the error is fatal.

In these cases I believe the error would be in the broker log?  `guest.output` should only be storing stdout/stderr from the actual exectuable?  Not from the shell?

**Comment by grondo** [comment]
> In these cases I believe the error would be in the broker log? guest.output should only be storing stdout/stderr from the actual exectuable? Not from the shell?

I'm talking about in the process after `fork(2)`. Any std output generated there should already be dup'd onto stdout/err of the job, right? (I could be wrong b/c I haven't looked at actual code)

It is much more useful to send errors from the child process to the stderr of the job, perhaps prefixed with `flux-shell` than sent back to job-exec module and the broker.log. This will be less of a concern when we send shell log messages to an output eventlog instead of stderr.

This is also how Slurm handles stdout/err after `fork(2)`, e.g. you get:

```console
$ srun foo
slurmstepd: error: execve(): foo: No such file or directory
srun: error: fluke3: task 0: Exited with exit code 2
```

Instead of just a report of a bad exit code with no context.

**Comment by grondo** [comment]
> Any std output generated there should already be dup'd onto stdout/err of the job, right? 

Ah, sorry I wasn't thinking clearly. Since exec() is wrapped in libsubprocess the shell doesn't have a way to emit any errors from the child process.

**Comment by grondo** [comment]
Actually I see libsubprocess `local_child()` can emit errors, but uses `flux_log_error()` on the parent's flux handle. I wonder if that is not appropriate and/or safe.

I'll try to open an issue

**Comment by chu11** [comment]
> I'm talking about in the process after fork(2). Any std output generated there should already be dup'd onto stdout/err of the job, right?

I see your point.  For the specific case here, an error on the `exec()` call leads to an error returned to the original caller of `flux_exec()`.  Do you think there should be something printed to stderr as well?

**Comment by chu11** [comment]
after implementing a solution where I emit "output-ready" after a subprocess has begun running, I realized there was a different issue.  I originally emit "output-ready" to indicate the "guest.output" directory has been created.  But if I emit "output-ready" after a subprocess has begun running, technically, I don't know of the "guest.output" directory has been created yet.  it's racy.

So I think PR #2374 is the right fix.  I'll re-order some commits so that we can push that PR first.

**Comment by chu11** [comment]
re-pushed and rebased on master.

**Comment by chu11** [comment]
re-pushed, noticed I had an unused variable

**Comment by garlick** [comment]
OK, this LGTM.  Thanks!
